### PR TITLE
Cpp17 ftw

### PR DESCRIPTION
**Pull request by suricactus**
Add support for cpp17 feature `for (auto [key, value]: asKeyValueRange(params)) {}`. All the credits to @m-kuhn

### CODE REVIEW

File: src/core/qfield.h [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#ifndef QFIELD_H
```
Comment (m-kuhn) [inline src/core/qfield.h:1]: License headers

File: src/core/qfield.h [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+#define QFIELD_H
+
+template <typename T>
+class asKeyValueRange
```
Comment (m-kuhn) [inline src/core/qfield.h:5]: Can we put this into a namespace (like `qgis.h`)?

Comment (m-kuhn) [reply src/core/qfield.h:5]: Oh, and a short "How to use" as docstring would be super cool

Comment (suricactus) [reply src/core/qfield.h]: Done

File: src/core/qfield.h (hunk 1 above)
Comment (suricactus) [reply src/core/qfield.h:18]: Aaah, sorry, should add this to the commit hooks :D

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1176-57328fd23c63b3c2390d7d7dffc1ef04bb1939b0-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1176-57328fd23c63b3c2390d7d7dffc1ef04bb1939b0-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1176-57328fd23c63b3c2390d7d7dffc1ef04bb1939b0-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1176-57328fd23c63b3c2390d7d7dffc1ef04bb1939b0-x86_64.apk)

File: CMakeLists.txt [hunk 3]
```diff
@@ -24,13 +24,13 @@ SET(CMAKE_AUTOUIC ON)
...
 
-SET(CMAKE_CXX_STANDARD 11)
+SET(CMAKE_CXX_STANDARD 17)
 SET(CMAKE_CXX_STANDARD_REQUIRED ON)
 
 ENABLE_TESTING()
 FIND_PACKAGE(QGIS)
 
-SET(QT_MIN_VERSION 5.13.0)
+SET(QT_MIN_VERSION 5.14.0)
```
Comment (m-kuhn) [inline CMakeLists.txt:33]: Is this a requirement?

File: src/core/qfield.h [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+
+/**
+ * Contains "polyfills" for backporting c++ features from standards > c++17 and Qt global methods
+ * added later than our minimum version, as well nice some sugar.
+ */
+namespace qfield
+{
+
+  template <typename T>
```
Comment (m-kuhn) [inline src/core/qfield.h:29]:
```suggestion
  /**
   * Can be used to iterate over `QMap`s in a convenient way.
   * 
   * \code{.cpp}
   * for (auto [key, value]: asKeyValueRange(map))
   * {
   *   // do something with `key` and `value`
   * }
   * \endcode
   */
  template <typename T>
```

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1176-11c564db327c96a43d2b61cec4efbb46d00cd8c4-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1176-11c564db327c96a43d2b61cec4efbb46d00cd8c4-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1176-11c564db327c96a43d2b61cec4efbb46d00cd8c4-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1176-11c564db327c96a43d2b61cec4efbb46d00cd8c4-x86_64.apk)

File: CMakeLists.txt (hunk 3 above)
Comment (suricactus) [reply CMakeLists.txt:33]: I just used this PR to fix this. After the recent changes in TextEdit widget validators, it is a requirement.

Comment (suricactus) [reply CMakeLists.txt:33]: https://github.com/opengisch/QField/blob/cpp17_ftw/src/qml/editorwidgets/TextEdit.qml#L1

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1176-858f5014555e960433634bcdc2acd670a01a302d-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1176-858f5014555e960433634bcdc2acd670a01a302d-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1176-858f5014555e960433634bcdc2acd670a01a302d-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1176-858f5014555e960433634bcdc2acd670a01a302d-x86_64.apk)

**Pull request merged**